docs(refactor): updates the cruise control rebalance concepts (#10810)
strimzi/strimzi-kafka-operator: documentation/modules/cruise-control/con-cruise-control-overview.adoc

//
// assembly-cruise-control-concepts.adoc

[id='con-cruise-control-overview-{context}']
= Cruise Control components and features

[role="_abstract"]
Cruise Control comprises four main components: 

Load Monitor:: Load Monitor collects the metrics and analyzes cluster workload data.
Analyzer:: Analyzer generates optimization proposals based on collected data and configured goals.
Anomaly Detector:: Anomaly Detector identifies and reports irregularities in cluster behavior.
Executor:: Executor applies approved optimization proposals to the cluster.

Cruise Control also provides a REST API for client interactions, which Strimzi uses to support these features:

* Generating optimization proposals from optimization goals
* Rebalancing a Kafka cluster based on an optimization proposal

NOTE: Other Cruise Control features are not currently supported, including self healing, notifications, and write-your-own goals.

== Optimization goals

Optimization goals define objectives for rebalancing, such as distributing topic replicas evenly across brokers. 

They are categorized as follows:

> * *Hard goals* are preset and mandatory for a proposal to succeed.

Reviewer: What do we mean by "mandatory" here? That the hard goals must be set? Or "must be satisfied"? The latter is a key distinction of hard goals.

Author: change to "must be satisfied"

> * *Soft goals* are optional and can be set aside if hard goals are met.

Reviewer: Even in the original doc, the phrase "They can be set aside if it means that all hard goals are met" sounds as if they soft goals are ignored or unimportant if hard goals are met. However, even if the hard goals are met, the soft goals are still important. Although the soft goals are best effort and will not prevent a optimization proposal from being generated, they are still taken it account when create a optimization proposal.

I would keep the original bullets concerning "hard goals" and "soft goals" but fix the line "They can be set aside if it means that all hard goals are met." to show that soft goals are "best effort"

Author: Thanks. Hopefully this is clearer to the intention:

* *Soft goals* are objectives that are prioritized during optimization as much as possible, without preventing a proposal from being created if all hard goals are satisfied.

* *Main goals* are inherited from Cruise Control, some preset as hard goals, used by default.

> * *Default goals* are the same as main goals by default but customizable.

Reviewer 2: As the name suggests, the Analyzer's `default.goals` is what it's used by default when requesting a proposal, and it is equal to the Analyzer's `goals` if the user does not set `default.goals`. 

The `goals` property is customizable in case the user wants to restrict the supported goals to a subset of what CruiseControl provides by default. This means that, if a goal is not present in Analyzer's `goals`, the it cannot be used in Analyzer's `default.goals` and proposal's `goals`.

The rest of the documentation seems to match my understanding. Do you think we can improve the wording a bit here?

Author: Thanks. I've tried to make this clearer and keep succinct.

> * *User-provided goals* are a subset of default goals configured for specific proposals.

Reviewer 2: "User provided" (here and in other places) is also a bit confusing for me, because all goals configurations can be user provided. What about `KafkaRebalance` goals?

Author: Thanks @fvaleri  
What about "proposal-specific"?  I think bringing a resource name into the list of types might be just as confusing.
@kyguy -- do you have any thoughts on this?

Reviewer: `proposal-specific` sounds good to me too

Configure optimization goals in the `Kafka` and `KafkaRebalance` custom resources.

* `Kafka` resource for default, hard, and main goals.
** Default goals: `Kafka.spec.cruiseControl.config.default.goals`
** Hard goals: `Kafka.spec.cruiseControl.config.hard.goals`
** Main goals: `Kafka.spec.cruiseControl.config.goals`
* `KafkaRebalance` resource for user-provided goals.
** User-Provided goals: `KafkaRebalance.spec.goals`

=== Hard and soft goals

> Hard goals are mandatory and must be satisfied for optimization proposals to be valid. 

Reviewer: ```suggestion
Hard goals are mandatory and must be satisfied for optimization proposals to be generated. 
```

Soft goals are best-effort objectives that Cruise Control tries to meet after all hard goals are satisfied.
The classification of hard and soft goals is fixed in Cruise Control code and cannot be changed.

A proposal meeting all hard goals is valid, even if it violates some soft goals. 

> Cruise Control prioritizes satisfying hard goals and then maximizes soft goals.

Reviewer: ```suggestion
Cruise Control prioritizes satisfying hard goals and then prioritizes satisfy soft goals in the order they are listed.
```

For example, a soft goal might be to evenly distribute a topic's replicas. 
Cruise Control will ignore this if it conflicts with hard goals.

Configure hard goals in your Cruise Control deployment using `Kafka.spec.cruiseControl.config.hard.goals`:

* To enforce all hard goals, omit the `hard.goals` property.
* To specify hard goals, list them in `hard.goals`.
* To exclude a hard goal, ensure it's not in either `default.goals` or `hard.goals`.

> Increasing the number of configured hard goals will reduce the likelihood of Cruise Control generating valid optimization proposals.

Reviewer: It'll reduce the likelihood of generating an optimization proposal period. From what I understand there is no concept of an invalid optimization proposal. If the hard goals cannot be satisfied, a proposal will not be generated
```suggestion
Increasing the number of configured hard goals will reduce the likelihood of Cruise Control generating optimization proposals.
```

You can approve or reject proposals and rerun them with adjusted goals if needed.

With Cruise Control deployed for use in Strimzi, the process to generate and approve an optimization proposal is as follows:

. Create a `KafkaRebalance` resource specifying optimization goals and any specific configurations. 
This resource triggers Cruise Control to initiate the optimization process.
. Strimzi Metrics Reporter runs in Kafka brokers, collecting raw metrics and publishing them to a dedicated Kafka topic (`strimzi.cruisecontrol.metrics`).
Metrics for partition topics and modeling the impact of rebalances are collected in other xref:proc-cruise-control-auto-created-topics-{context}[topics automatically created when Cruise Control is deployed]. 
. Load Monitor collects the metrics from Kafka brokers, including CPU, disk, and network utilization data.
. Anomaly Detector continuously monitors the collected metrics to identify anomalies, such as broker failures or disk capacity issues, that could impact cluster stability.
. Analyzer processes the collected metrics and constructs a _workload model_ of the current state of the Kafka cluster. 

> It generates an optimization proposal based on configured goals (and constraints) such as balancing partition distribution across brokers, which is sent to the status of the `KafkaRebalance` resource.

Reviewer 2: ```suggestion
Based on configured goals and capacities, it generates an optimization proposal for balancing partition distribution across brokers, which is finally reflected in the status of the `KafkaRebalance` resource.
```

Approving an optimization proposal:: You approve the optimization proposal by setting the `strimzi.io/rebalance` annotation of the `KafkaRebalance` resource to `approve`.
Cruise Control applies the proposal to the Kafka cluster and starts a cluster rebalance operation.
Rejecting an optimization proposal:: If you choose not to approve an optimization proposal,
you can xref:proc-generating-optimization-proposals-str[change the optimization goals] or xref:con-rebalance-{context}[update any of the rebalance performance tuning options], and then generate another proposal.
You can generate a new optimization proposal for a `KafkaRebalance` resource by setting the `strimzi.io/rebalance` annotation to `refresh`.

Use optimization proposals to assess the movements required for a rebalance.
For example, a summary describes inter-broker and intra-broker movements.
Inter-broker rebalancing moves data between separate brokers.
Intra-broker rebalancing moves data between disks on the same broker when you are using a JBOD storage configuration.
Such information can be useful even if you don't go ahead and approve the proposal.

> You might reject an optimization proposal, or delay its approval, because of the additional load on a Kafka cluster when rebalancing.

Reviewer 2: We might want to add that if the proposal is too old, the cluster load may have changed significantly, so it would be better to request a new proposal.

Fewer data movements at a single time reduces the load on the cluster.

.Example optimization proposal summary
[source,yaml]
----
Name:         my-rebalance
Namespace:    myproject
Labels:       strimzi.io/cluster=my-cluster
Annotations:  API Version:  kafka.strimzi.io/v1alpha1
Kind:         KafkaRebalance
Metadata:
# ...
Status:
  Conditions:
    Last Transition Time:  2022-04-05T14:36:11.900Z
    Status:                ProposalReady
    Type:                  State
  Observed Generation:     1
  Optimization Result:
    Data To Move MB:  0
    Excluded Brokers For Leadership:
    Excluded Brokers For Replica Move:
    Excluded Topics:
    Intra Broker Data To Move MB:         12
    Monitored Partitions Percentage:      100
    Num Intra Broker Replica Movements:   0
    Num Leader Movements:                 24
    Num Replica Movements:                55
    On Demand Balancedness Score After:   82.91290759174306
    On Demand Balancedness Score Before:  78.01176356230222
    Recent Windows:                       5
  Session Id:                             a4f833bd-2055-4213-bfdd-ad21f95bf184
----

The proposal will also move 24 partition leaders to different brokers.

> This requires a change to the ZooKeeper configuration, which has a low impact on performance.

Reviewer 2: ```suggestion
This requires a change to the cluster metadata, which has a low impact on performance.
```

Cruise Control first prioritizes satisfying hard goals, and then addresses soft goals in the order they are listed.
A proposal meeting all hard goals is valid, even if it violates some soft goals. 

For example, a soft goal might be to evenly distribute a topic's replicas. 
Cruise Control will ignore this if it conflicts with hard goals.

Configure hard goals in your Cruise Control deployment using `Kafka.spec.cruiseControl.config.hard.goals`:

* To enforce all hard goals, omit the `hard.goals` property.
* To specify hard goals, list them in `hard.goals`.
* To exclude a hard goal, ensure it's not in either `default.goals` or `hard.goals`.

Increasing the number of configured hard goals will reduce the likelihood of Cruise Control generating optimization proposals.

=== Main goals

Main goals are predefined and available to all users in Cruise Control. 
Goals not listed as main goals cannot be used in Cruise Control operations.
Some main goals are preset as hard goals.

To simplify configuration, use the inherited main goals unless you need to exclude specific goals from `KafkaRebalance` resources. 

> You can adjust the priority order in the default optimization goals configuration.

Reviewer: To my understanding yes, the goals are listed in order of priority, so the first goal listed in the configuration will have higher priority than the second goal list

> You can adjust the priority order in the default optimization goals configuration.

Reviewer 3: Ok. So I think we should clarify here (and everywhere goals priority is mentioned) that the priority is based on the order. Not sure if "priority order" is used for this purpose but it's not clear to me.

> You can adjust the priority order in the default optimization goals configuration.

Reviewer 2: We have to admit that CC naming doesn't help much here.

IMO, this is the default inter-broker goals priority list:

https://github.com/linkedin/cruise-control/blob/main/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/AnalyzerConfig.java#L312-L327

To change it, the user can configure Kafka `default.goals` or KafkaRebalance `goals` with a different order. Only goals listed in Kafka `goals` can be used in the previous configuration (DEFAULT_DEFAULT_GOALS is a subset of DEFAULT_GOALS).

- Do you agree?
- Are we exposing Kafka `goals` because we intend to eventually support custom goals?

=== Rebalancing modes

Proposals for rebalances can be generated in three modes, which are specified using the `spec.mode` property of the `KafkaRebalance` custom resource.

`full` mode:: The `full` mode runs a full rebalance by moving replicas across all the brokers in the cluster.
This is the default mode if the `spec.mode` property is not defined in the `KafkaRebalance` custom resource.

`add-brokers` mode:: The `add-brokers` mode is used after scaling up a Kafka cluster by adding one or more brokers.
Normally, after scaling up a Kafka cluster, new brokers are used to host only the partitions of newly created topics.
If no new topics are created, the newly added brokers are not used and the existing brokers remain under the same load.
By using the `add-brokers` mode immediately after adding brokers to the cluster, the rebalancing operation moves replicas from existing brokers to the newly added brokers.
You specify the new brokers as a list using the `spec.brokers` property of the `KafkaRebalance` custom resource.

`remove-brokers` mode:: The `remove-brokers` mode is used before scaling down a Kafka cluster by removing one or more brokers.

> If you scale down a Kafka cluster, brokers are shut down even if they host replicas.

Reviewer 3: this is actually not the case by default. We have now I check which prevents you to scale down if the brokers are hosting partitions, so that you need to run rebalancing first. There is a `strimzi.io/skip-broker-scaledown-check` annotation which can be set to true to skip this check (but it's disabled by default as just explained).

Author: Right. I've updated this and added a note that references the section on skipping checks on scale-down operations

A summary of a successfully generated optimization proposal is contained in the `Status.optimizationResult` property of the `KafkaRebalance` resource.
The information provided is a summary of the full optimization proposal.
Broker load:: The broker load is stored in a ConfigMap that contains data as a JSON string. The broker load shows before and after values for the proposed rebalance, so you can see the impact on each of the brokers in the cluster.

=== Manually approving or rejecting an optimization proposal

An optimization proposal summary shows the proposed scope of changes.

You can use the name of the `KafkaRebalance` resource to return a summary from the command line.

.Returning an optimization proposal summary
[source,shell]
----
kubectl describe kafkarebalance <kafka_rebalance_resource_name> -n <namespace>
----

You can also use the `jq` {JQTool}.

.Returning an optimization proposal result using jq
[source,shell]
----
kubectl get kafkarebalance <kafka_rebalance_resource_name> -n <namespace> -o json | jq '.status.optimizationResult'
----

Use the summary to decide whether to approve or reject an optimization proposal.

> Approving an optimization proposal:: You approve the optimization proposal by setting the `strimzi.io/rebalance` annotation of the `KafkaRebalance` resource to `approve`.

Reviewer: Looks like there is a section `Approving optimization proposals`, should this be referenced and/or moved closer to auto approval section?

//
// assembly-cruise-control-concepts.adoc

[id='con-cruise-control-overview-{context}']
= Cruise Control components and features

[role="_abstract"]
Cruise Control comprises four main components: 

Load Monitor:: Load Monitor collects the metrics and analyzes cluster workload data.
Analyzer:: Analyzer generates optimization proposals based on collected data and configured goals.
Anomaly Detector:: Anomaly Detector identifies and reports irregularities in cluster behavior.
Executor:: Executor applies approved optimization proposals to the cluster.

Cruise Control also provides a REST API for client interactions, which Strimzi uses to support these features:

* Generating optimization proposals from optimization goals
* Rebalancing a Kafka cluster based on an optimization proposal
* Changing topic replication factor

NOTE: Other Cruise Control features are not currently supported, including self healing, notifications, and write-your-own goals.

== Optimization goals

Optimization goals define objectives for rebalancing, such as distributing topic replicas evenly across brokers. 

They are categorized as follows:

> * *Main goals* are a list of goals supported by the Cruise Control instance and used in its operations. 

Reviewer: ```suggestion
* *Main goals* are a list of goals supported by the Cruise Control instance that can be used in its operations. 
```

Reviewer 3: I agree, not sure that "main" makes much sense

Author: Yeah. Supported is much better. Changed.

> By default, this list includes all built-in goals. 

Reviewer: What do we mean by "built-in" here? All the goals that come included/shipped with upstream Cruise Control?

For a goal to be used in other categories, such as default or hard goals, it must first be listed in main goals. 
To prevent a goal’s usage, remove it from this list.
* *Hard goals* are preset and must be satisfied for a proposal to succeed.

> * *Soft goals* are objectives that are prioritized during optimization as much as possible, without preventing a proposal from being created if all hard goals are satisfied.

Reviewer: ```suggestion
* *Soft goals* are objectives that are prioritized during optimization as much as possible, but won't prevent the generation of a proposal if not satisfied. 
```

Reviewer: IIRC, at this time, both hard goals and soft goals are preset by upstream Cruise Control. Basically, whatever is not a `hard.goal` (whatever `goals` not listed in the default `hard.goals` config) are soft goals. 

I am not completely sure why Cruise Control offers this `hard.goals` list in the first place if the goal type is hard coded/preset by upstream Cruise Control. Listing a goal in the `hard.goals` list does not change it into a hard goal from what I remember (but my memory is hazy so let's leave this information out for now, I just wanted people to be aware of it)

I bring this up because it may be worth mentioning that soft goals are preset as well

* *Default goals* refer to the goals used by default when generating proposals. 
They match the main goals unless specifically set by the user.
* *Proposal-specific goals* are a subset of default goals configured for specific proposals.

Configure optimization goals in the `Kafka` and `KafkaRebalance` custom resources.

* `Kafka` resource for main, hard, and default goals.
** Main goals: `Kafka.spec.cruiseControl.config.goals`
** Hard goals: `Kafka.spec.cruiseControl.config.hard.goals`
** Default goals: `Kafka.spec.cruiseControl.config.default.goals`
* `KafkaRebalance` resource for proposal-specific goals.
** Proposal-specific goals: `KafkaRebalance.spec.goals`

=== Main goals

> Main goals are predefined and available to all users in Cruise Control. 

Reviewer: ```suggestion
Main goals are predefined and available to use for generating Cruise Control optimization proposals. 
```

Supported goals are predefined and available to use for generating Cruise Control optimization proposals. 
Goals not listed as supported goals cannot be used in Cruise Control operations.
Some supported goals are preset as hard goals.

> To simplify configuration, use the inherited supported goals unless you need to exclude specific goals from `KafkaRebalance` resources. 

Reviewer: This  seems to imply that users should edit the `goals` list if they want to exclude goals from the `KafkaRebalance` resource. I would remove this line

> To simplify configuration, use the inherited supported goals unless you need to exclude specific goals from `KafkaRebalance` resources. 

Reviewer: This line seems to imply that if we don't want specific inherited supported goals from being used by `KafkaRebalance` resources we can update the supported goal list. However, this could potentially cause problems with the the `default.goals` list and other lists. To "exclude specific goals from `KafkaRebalance` resources"  we would just remove the goal from the `KafkaRebalance` resource itself. To make sure exclude a specific goal from being used in any goal list in Cruise Control we would remove it from the supported goals list

I think we should either remove this sentence or change it to something like:

```suggestion
To simplify configuration, use the inherited supported goals.
However, if you don't want specific goals to be supported or used for any proposal generation generation, remove them from the supported goals list. 
```
That being said, I think this  information is already covered by the first two sentences of this section, so I am for removing this line

Author: I think its simpler to remove. Removed.

> You can adjust the priority order in the default optimization goals configuration.

Reviewer: Since this is about the `default.goals` list, maybe we should move this line to the "Default goals" section

Configure main goals in `Kafka.spec.cruiseControl.config.goals`:

* To accept inherited main goals, omit the `goals` property.
* To modify main goals, specify the goals in descending priority order in the `goals` property.

=== Hard and soft goals

> Hard goals are mandatory and must be satisfied for optimization proposals to be generated. 

Reviewer: What do we mean by "mandatory" here?

Author: I've removed that.

Soft goals are best-effort objectives that Cruise Control tries to meet after all hard goals are satisfied.
The classification of hard and soft goals is fixed in Cruise Control code and cannot be changed.

Cruise Control first prioritizes satisfying hard goals, and then addresses soft goals in the order they are listed.
A proposal meeting all hard goals is valid, even if it violates some soft goals. 

For example, a soft goal might be to evenly distribute a topic's replicas. 

> Cruise Control will ignore this if it conflicts with hard goals.

Reviewer: I wouldn't say the Cruise Control will ignore the soft goal, it just continue to generate an optimization proposal even if the soft goal isn't completely satisfied

Configure hard goals in your Cruise Control deployment using `Kafka.spec.cruiseControl.config.hard.goals`:

* To enforce all hard goals, omit the `hard.goals` property.
* To specify hard goals, list them in `hard.goals`.
* To exclude a hard goal, ensure it's not in either `default.goals` or `hard.goals`.

Increasing the number of configured hard goals will reduce the likelihood of Cruise Control generating optimization proposals.

=== Default goals

Cruise Control uses default goals to generate an optimization proposal. 
You can override default goals by setting proposal-specific optimization goals in a `KafkaRebalance` resource.

> If `default.goals` is not specified in the Cruise Control deployment configuration, main goals are used as default goals. 

Reviewer: ```suggestion
If `default.goals` is not specified in the Cruise Control deployment configuration, Strimzi will configure the `default.goals` to the list of goals specified in `main.goals`. 
```

Author: I think we mean `config.goals` not `main.goals`?

> The optimization proposal is then generated using these main goals.

Reviewer: ```suggestion
The optimization proposal based on this goals list will then be generated and cached.
```

> If `default.goals` is not specified in the Cruise Control deployment configuration, Strimzi configures `default.goals` to the list of goals specified in `config.goals`. 

Reviewer: I believe the CC configuration is just `goals` although this field is in the `cruiseControl.config` section of the `Kafka` resource, just like `default.goals` is

```suggestion
If `default.goals` is not specified in the Cruise Control deployment configuration, Strimzi configures `default.goals` to the list of supported goals specified in `goals`. 
```

> The optimization proposal based on this goals list is then generated and cached..

Reviewer: ```suggestion
The optimization proposal based on this supported goals list is then generated and cached..
```

* To use main goals as default, omit the `default.goals` property.
* To modify default goals, specify a subset of main goals in the `default.goals` property.

=== Proposal-specific goals

> Proposal-specific optimization goals narrow down the configured default goals for specific optimization proposals.

Reviewer: Not necessarily "narrows down", the configured default goals list could be smaller than proposal-sepcific goals list.
```suggestion
Proposal-specific optimization goals allow creating optimization proposals based on a specific list of goals.
```

Proposal-specific optimization goals support the creation of optimization proposals based on a specific list of goals.

Configure proposal-specific goals in `KafkaRebalance.spec.goals`, specifying a subset of supported optimization goals for customization.

For example, you can optimize topic leader replica distribution across the Kafka cluster without considering disk capacity or utilization by defining a single proposal-specific goal.

=== Goals order of priority

Unless you change the Cruise Control xref:proc-configuring-deploying-cruise-control-{context}[deployment configuration], Strimzi inherits goals from Cruise Control, in descending priority order.

The following list shows supported goals inherited by Strimzi from Cruise Control in descending priority order.
Goals labeled as hard are mandatory constraints that must be satisfied for optimization proposals.

* `RackAwareGoal` (hard)

> * `MinTopicLeadersPerBrokerGoal`

Reviewer: ```suggestion
* `MinTopicLeadersPerBrokerGoal` (hard)
```

Hard goals must be included as a subset in the supported, default, and proposal-specific goals.

.Example `KafkaRebalance` configuration for proposal-specific goals
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaRebalanceApiVersion}
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
spec:
  goals:
    - RackAwareGoal
    - TopicReplicaDistributionGoal
  skipHardGoalCheck: true
----

=== Skipping hard goal checks

If `skipHardGoalCheck: true` is specified in the `KafkaRebalance` custom resource, Cruise Control does not verify that the proposal-specific goals include all the configured hard goals. 
This allows for more flexibility in generating optimization proposals, but may lead to proposals that do not satisfy all hard goals.

However, any hard goals included in the proposal-specific goals will still be treated as hard goals by Cruise Control, even with `skipHardGoalCheck: true`.

[id='con-optimization-proposals-{context}']

> == Optimization proposals

Reviewer: What's the motivation behind providing summary of how the internal components of Cruise Control work? Are there things we want to mention here that are not already mentioned in the upstream documentation? [1]

I agree with providing clarification/documentation surrounding to points that are Strimzi specific or unclear from upstream Cruise Control documentation. However, it seems some of the sentences here summarize Cruise Control components whose function is already documented/described in the upstream Cruise Control docs [1]. i.e the sentences surrounding the Load Monitor, Analyzer, and Anomaly Detector.

[1] https://github.com/linkedin/cruise-control/wiki/Overview

Optimization proposals are summaries of proposed changes based on the defined optimization goals, assessed in a specific order of priority. 
You can approve or reject proposals and rerun them with adjusted goals if needed.

With Cruise Control deployed for use in Strimzi, the process to generate and approve an optimization proposal is as follows:

. Create a `KafkaRebalance` resource specifying optimization goals and any specific configurations. 

> This resource triggers Cruise Control to initiate the optimization process.

Reviewer: ```suggestion
This resource triggers Cruise Control to initiate the optimization proposal generation process.
```

> . Strimzi Metrics Reporter runs in Kafka brokers, collecting raw metrics and publishing them to a dedicated Kafka topic (`strimzi.cruisecontrol.metrics`).

Reviewer: The Strimzi Metrics Reporter is a different metrics reporter
```suggestion
. Cruise Control Metrics Reporter runs in Kafka brokers, collecting raw metrics and publishing them to a dedicated Kafka topic (`strimzi.cruisecontrol.metrics`).
```

> Metrics for partition topics and modeling the impact of rebalances are collected in other xref:proc-cruise-control-auto-created-topics-{context}[topics automatically created when Cruise Control is deployed]. 

Reviewer: ```suggestion
Metrics for brokers, topics, and partitions are aggregated, sampled and then stored in other xref:proc-cruise-control-auto-created-topics-{context}[topics automatically created when Cruise Control is deployed]. 
```

> . Load Monitor collects the metrics from Kafka brokers, including CPU, disk, and network utilization data, and sends load data to the Analyzer and Anomaly Detector.

Reviewer: The load monitor is the component collecting, processing and storing the metrics mentioned in the previous sentence. It creates a cluster model of the load of the cluster which is used by the Analyzer and Anomaly Detector

. Anomaly Detector continuously monitors the health and performance of the Kafka cluster, checking for things like broker failures or disk capacity issues, that could impact cluster stability.

> . Analyzer processes the collected metrics and anomalies to construct a _workload model_ of the current state of the Kafka cluster. 

Reviewer: The analyzer creates optimization proposals based on the workload models from the Load Monitor

> Based on configured goals and capacities, it generates an optimization proposal for balancing partition distribution across brokers.

Reviewer: I am not sure I understand the phrase "balancing partition distributions"? Is this a typo? 

Should it say "balancing partitions across brokers" or "evenly distributing partitions across brokers", or something along those lines?

Through the REST API, a summary of the proposal is reflected in the status of the `KafkaRebalance` resource.
. The optimization proposal is approved or rejected (manually or automatically) based on its alignment with cluster management goals.
. If approved, the Executor applies the optimization proposal to rebalance the Kafka cluster. 
This involves reassigning partitions and redistributing workload across brokers according to the approved proposal.

.Cruise Control optimization process
image:kafka-concepts-cruise-control.png[Cruise Control process]

> Optimization proposals comprise separate partition reassignment commands.

Reviewer: ```suggestion
Optimization proposals comprise of a list of partition reassignment mappings.
```

> When you approve a proposal, the Cruise Control server applies these commands to the Kafka cluster.

Reviewer: ```suggestion
When you approve a proposal, the Cruise Control server applies these reassignments to the Kafka cluster.
```

* *Hard goals* are preset and mandatory for a proposal to succeed.
* *Soft goals* are optional and can be set aside if hard goals are met.
* *Main goals* are inherited from Cruise Control and typically used as the default set of goals, with some preset as hard goals.

> The goals can be customized to specify a subset of all available goals provided by Cruise Control.

Reviewer 2: What about:

```
* *Main goals* contains the list of goals that can be used by the Cruise Control operations (all built-in goals by default).
This list can be customized to specify a subset of all the available goals, which may include custom goals.
```

My intention is to avoid giving the impression that you should update this configuration to change default goals. You can, but it would be an indirect update, and your configuration may be lost if someones arrives and sets `default.goals`.
